### PR TITLE
Plugin System

### PR DESCRIPTION
**Pull request by rafaelvieiras**
### Explanation of the changes, problem that they are intended to fix.
On this implementation, I create a base for plugin system.

For now, 3 hooks are available to use on plugins:
- `dizquetv:WEB:CustomCss`
- `dizquetv:WEB:CustomJs`
- `dizquetv:WEB:RegisterSettings`

I create a plugin for example, this is a Dark Mode plugin. See the [repository on here](https://github.com/rafaelvieiras/dizquetv-plugin-darkmode).
This plugin is listed on `src/plugins/plugins.json`, so it's can be installed using UI.

I'm working on a documentation, and it's are [been building here](https://github.com/vexorian/dizquetv/wiki/Plugins).

After this PR, my next steps is to work into hooks to make possible creation of an Authentication Plugin, to make dizqueTV most secure and decouple cache from dizqueTV to create a plugin with this propose.

### All Submissions:

* [x] I have read the code of conduct.
* [x] I am submitting to the correct base branch

### New features

* [x] I understand that the feature may not be accepted if it doesn't fit the upstream app's planned design direction. But that in this case I am encouraged to share this as an available modification other users can use if they want.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -71,27 +75,41 @@ if(!fs.existsSync(path.join(process.env.DATABASE, 'cache'))) {
...
+}
 
 
 channelDB = new ChannelDB( path.join(process.env.DATABASE, 'channels') );
 fillerDB = new FillerDB( path.join(process.env.DATABASE, 'filler') , channelDB, channelCache );
 
 customShowDB = new CustomShowDB( path.join(process.env.DATABASE, 'custom-shows') );
 
-db.connect(process.env.DATABASE, ['channels', 'plex-servers', 'ffmpeg-settings', 'plex-settings', 'xmltv-settings', 'hdhr-settings', 'db-version', 'client-id', 'cache-images', 'settings'])
+db.connect(process.env.DATABASE, ['channels', 'plex-servers', 'ffmpeg-settings', 'plex-settings', 'xmltv-settings', 'hdhr-settings', 'db-version', 'client-id', 'cache-images', 'settings', 'plugin-manager'])
```
Comment (vexorian) [inline index.js:88]: I have multiple concerns here. The method to rely on storage to detect reboots, seems odd. Then we have the usage of the settings db, which really shouldn't be used for that. Finally, I am not sure why do we need this hook? Since dizquetv is always running, then you can use a server start hook to cover all of these use cases (when the server starts, you can basically assume it has been restarted)

If you really do have ideas that require a restart trigger specifically that wouldn't work with a start trigger, then tell me.

File: web/directives/toast-notifications.js [hunk 2]
```diff
@@ -98,6 +98,10 @@ module.exports = function ($timeout) {
                 let normalEvent = (title) => {
                     return (event) => {
                         let data = JSON.parse(event.data);
+
+                        const customEvent = new CustomEvent('PushEvent', { detail: data.detail });
+                        window.dispatchEvent(customEvent);
```
Comment (vexorian) [inline web/directives/toast-notifications.js:103]: What's this supposed to do? It seems unrelated to plugins

File: index.js (hunk 1 above)
Comment (rafaelvieiras) [reply index.js:88]: Let me contextualize you.

We need to dizquetv be restarted after an installation or uninstallation of plugins.

Because plugin system use node `require()` to import all plugins, after a plugin is removed, they stay on node memory and broken the server.

My initial Idea, are to make the server restart, but, to make this, I need to create a script to start before the server (express), and store or manager server process.

In the future, we need to make this, for a good practice and redundancy, but for now, to make this is required a big refac on index.js

Look to fix this problem, in a fast way, I created a boolean setting `serverRestart`, to be true when a plugin is installed/unstalled.
With this variable, I respond on router getNeedRestart if serve need to be restarted or not.

This router is called from frontend. When the value is true, the frontend show an alert with a message to user to make a restart on server.
image

I can't find a way to store this variable to be changed and used from 3 different scripts, without create a singleton to only this. And I don't know if create a singleton, just for this, just for 1 variable, is a good practice.

I thought it might be more reliable to store this information in the DB.

If you believe this is a bad way, You can help me find another way? 😄

File: web/directives/toast-notifications.js (hunk 2 above)
Comment (rafaelvieiras) [reply web/directives/toast-notifications.js:103]: This event is a frontend only event, this sends push's data to all `EventListener` window on channel `PushEvent`.

With this custom event, all web plugins can receive push information on a isolate js scope and can communicate with another plugins without the backend, it's a "modern version" of postMessage 🤢, but better, lighter, for current window scope (not for wall browser windows) and recommended instead of postMessage.

For now, we use this only for `plugin.uninstall` and `plugin.install` events sent from backend over EventSource. With this events on `web/app.js` we reload page and store restart information on web scope to show restart alert.

👍🏻

**Comment by vexorian** [comment]
Good Morning.

I am really sorry that the next thing I will do is postpone  consideration of your PR for a later version. I am specially sorry that it took me so long to reply.

Although these last weekends have certainly been very busy/tired ones and I didn't have a lot of time I could use in dizqueTV, the truth is that I've been procrastinating on my upstream work for dizqueTV and last night I realized it's because making a decision on the plugin system PR is the next thing on my queue and I am not sure if I want this?

There's some reasoning for this.

* dTV is still a very young project and I am not sure how much of the code is going to stay. It becomes a problem if I do a massive change and it breaks plugins. Communities have gotten really mad about the loss of plugins in the past.
* I think I like the current style of customization. Letting users edit files like the CSS or images and then also having the abilities to do automation with the rest api. I think that for the short to middle term it is a better idea to improve things on this area instead of relying on plugins (and one benefit is that these automations don't have to stick to a specific programming language). I think by adding more sse endpoints we can allow some way to have the relevant hooks.
* For ffmpeg maybe we can figure out a way to customize the filters/command line better?

That said, I am still undecided. I'll leave the PR open in case other people have different opinions about whether or not we need a plugin system and maybe in a later version the need for it becomes clearer.

As usual you are free to maintain a fork and if anyone finds need for plugins they may use it.

**Comment by rafaelvieiras** [comment]
I close this PR because now I focus on https://github.com/ErsatzTV 
Anyone is free to get this branch and continue the work. 👍🏻
